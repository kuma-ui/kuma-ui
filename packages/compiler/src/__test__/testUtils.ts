import { compile } from "../compile";
import { compileSync } from "..";

type Result = ReturnType<typeof compile>;

export function getExpectSnapshot(result: Result) {
  return `
  ${result.css}
  
  ${result.code}
  `;
}

type CompileResult = ReturnType<typeof compileSync>;

export function getExpectSnapshotSync(result: CompileResult) {
  return `
  ${result?.css}
  
  ${result?.code}
  `;
}
