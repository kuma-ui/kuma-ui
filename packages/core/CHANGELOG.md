# @kuma-ui/core

## 1.5.6

### Patch Changes

- [#398](https://github.com/kuma-ui/kuma-ui/pull/398) [`08632b1`](https://github.com/kuma-ui/kuma-ui/commit/08632b17471b62da3d2e3c0d6fec0097ee1cd514) Thanks [@yukukotani](https://github.com/yukukotani)! - Fix `DOMException: Failed to execute 'deleteRule' on 'CSSStyleSheet': The index provided (4294967295) is larger than the maximum index (0).`

- Updated dependencies [[`b3490c1`](https://github.com/kuma-ui/kuma-ui/commit/b3490c10c30c9fa65167bb51554903397f4a90a5), [`a0dc9cd`](https://github.com/kuma-ui/kuma-ui/commit/a0dc9cda8ecfad419be8053c1237d4d6551ef77b)]:
  - @kuma-ui/system@1.7.3

## 1.5.5

### Patch Changes

- [#385](https://github.com/kuma-ui/kuma-ui/pull/385) [`407a55e`](https://github.com/kuma-ui/kuma-ui/commit/407a55e4b2e4f814fd4a3ea3539044bb2d5d0937) Thanks [@kotarella1110](https://github.com/kotarella1110)! - refactor: improve toCssUnit logic to reduce bundle size

- Updated dependencies [[`407a55e`](https://github.com/kuma-ui/kuma-ui/commit/407a55e4b2e4f814fd4a3ea3539044bb2d5d0937), [`2bb1f6d`](https://github.com/kuma-ui/kuma-ui/commit/2bb1f6d449f87d5fab6a1fd23cd1a2f925478c7d)]:
  - @kuma-ui/system@1.7.2

## 1.5.4

### Patch Changes

- [#380](https://github.com/kuma-ui/kuma-ui/pull/380) [`d2fe302`](https://github.com/kuma-ui/kuma-ui/commit/d2fe302a0d841c19cd628a94e12423a89599f3cc) Thanks [@yukukotani](https://github.com/yukukotani)! - avoid false-positive hasDynamicProps

- Updated dependencies [[`afd4d57`](https://github.com/kuma-ui/kuma-ui/commit/afd4d579af008fcd9fd8f602956ff665590ffca5)]:
  - @kuma-ui/system@1.7.1

## 1.5.3

### Patch Changes

- [#372](https://github.com/kuma-ui/kuma-ui/pull/372) [`3840e83`](https://github.com/kuma-ui/kuma-ui/commit/3840e835ed3c3c6bb707a565ecec91f2e737eb18) Thanks [@AdePhil](https://github.com/AdePhil)! - chore: Refactor ts-pattern to use switch/if-else

- Updated dependencies [[`d439b72`](https://github.com/kuma-ui/kuma-ui/commit/d439b72e9dd651ccd65a889c1dce5b3fc4e54a11), [`aa23d92`](https://github.com/kuma-ui/kuma-ui/commit/aa23d92f93b0373b91c7b0d431a5745aa06380b9)]:
  - @kuma-ui/system@1.7.0

## 1.5.2

### Patch Changes

- Updated dependencies [[`b5415ed`](https://github.com/kuma-ui/kuma-ui/commit/b5415ed684c5212a4dd6d6e10d43969164375233)]:
  - @kuma-ui/system@1.6.0

## 1.5.1

### Patch Changes

- [#311](https://github.com/kuma-ui/kuma-ui/pull/311) [`69b90fa`](https://github.com/kuma-ui/kuma-ui/commit/69b90fadedd338d552aa6c128226ea7732f7fafa) Thanks [@poteboy](https://github.com/poteboy)! - Update prettier from 2.8 to 3.0.3

- [#351](https://github.com/kuma-ui/kuma-ui/pull/351) [`e8e5195`](https://github.com/kuma-ui/kuma-ui/commit/e8e5195706f90e99a0236d76c5825e5060425b7a) Thanks [@yukukotani](https://github.com/yukukotani)! - Override original prop types with `as` component props

- Updated dependencies [[`d8b82dc`](https://github.com/kuma-ui/kuma-ui/commit/d8b82dc7e851afa994eda664e09acff41d04ba6c)]:
  - @kuma-ui/system@1.5.2

## 1.5.0

### Minor Changes

- [#343](https://github.com/kuma-ui/kuma-ui/pull/343) [`1e31d15`](https://github.com/kuma-ui/kuma-ui/commit/1e31d159d968e87054bb64155be1618d568be172) Thanks [@poteboy](https://github.com/poteboy)! - Implement ForwardRef in Components

- [#342](https://github.com/kuma-ui/kuma-ui/pull/342) [`2c3da38`](https://github.com/kuma-ui/kuma-ui/commit/2c3da38efccea069e9a4ac8e4fa7f5e7d50d0db6) Thanks [@mmkal](https://github.com/mmkal)! - styled-components syntax parity

## 1.4.1

### Patch Changes

- Updated dependencies [[`6bfcab7`](https://github.com/kuma-ui/kuma-ui/commit/6bfcab7b85de6ff020dec3e65cf936aebd14be54)]:
  - @kuma-ui/system@1.5.1

## 1.4.0

### Minor Changes

- [#240](https://github.com/kuma-ui/kuma-ui/pull/240) [`4073398`](https://github.com/kuma-ui/kuma-ui/commit/4073398d58e011f0afb351025e9b326ff56c0e36) Thanks [@poteboy](https://github.com/poteboy)! - Enhance Performance by Optimizing Compiler for Static HTML Conversion

- [#321](https://github.com/kuma-ui/kuma-ui/pull/321) [`8e1a211`](https://github.com/kuma-ui/kuma-ui/commit/8e1a211860281f2fb7eb94bb2f8680300ebf7b7c) Thanks [@yukukotani](https://github.com/yukukotani)! - Introduce `defaultProps` in component theme

### Patch Changes

- [#304](https://github.com/kuma-ui/kuma-ui/pull/304) [`1b7982e`](https://github.com/kuma-ui/kuma-ui/commit/1b7982e5d8b9ab4f862d1ff113353380b411129e) Thanks [@yukukotani](https://github.com/yukukotani)! - Improve autocompletion experience of `baseStyle` and `variants`

- [#324](https://github.com/kuma-ui/kuma-ui/pull/324) [`0fc8f4d`](https://github.com/kuma-ui/kuma-ui/commit/0fc8f4d86972517052aacb982fb9c9d94788353a) Thanks [@kotarella1110](https://github.com/kotarella1110)! - improve createTheme type

- [#316](https://github.com/kuma-ui/kuma-ui/pull/316) [`f441c1d`](https://github.com/kuma-ui/kuma-ui/commit/f441c1d0b3a5e2e13d83aeff4984ad64702b69d9) Thanks [@yukukotani](https://github.com/yukukotani)! - Exclude vitest.setup.ts from dist files

- Updated dependencies [[`8d68550`](https://github.com/kuma-ui/kuma-ui/commit/8d685505e17beceafd43d9d3618342e1246032e4), [`031bc85`](https://github.com/kuma-ui/kuma-ui/commit/031bc851c140d5fbbd7a6b40116c2b894113940e), [`8e1a211`](https://github.com/kuma-ui/kuma-ui/commit/8e1a211860281f2fb7eb94bb2f8680300ebf7b7c)]:
  - @kuma-ui/system@1.5.0
  - @kuma-ui/sheet@1.3.0

## 1.3.0

### Minor Changes

- [#293](https://github.com/kuma-ui/kuma-ui/pull/293) [`638d61d`](https://github.com/kuma-ui/kuma-ui/commit/638d61df0a6dd355d8bb4847e5752877a9895d71) Thanks [@yukukotani](https://github.com/yukukotani)! - Add support for z-index token

### Patch Changes

- Updated dependencies [[`638d61d`](https://github.com/kuma-ui/kuma-ui/commit/638d61df0a6dd355d8bb4847e5752877a9895d71), [`3414300`](https://github.com/kuma-ui/kuma-ui/commit/341430056cff0e4097e0b066726e5c110219eac9), [`314096c`](https://github.com/kuma-ui/kuma-ui/commit/314096cde1d45ffa35c93ef49bddacb7e745606b)]:
  - @kuma-ui/system@1.4.0
  - @kuma-ui/sheet@1.2.0

## 1.2.0

### Minor Changes

- [#274](https://github.com/kuma-ui/kuma-ui/pull/274) [`bc1a7f9`](https://github.com/kuma-ui/kuma-ui/commit/bc1a7f979f5539a4e2d1d2785dc2a893eee076eb) Thanks [@yukukotani](https://github.com/yukukotani)! - Add support for spacing theme token

- [#289](https://github.com/kuma-ui/kuma-ui/pull/289) [`10f9811`](https://github.com/kuma-ui/kuma-ui/commit/10f98111d542f1fa45a6cda18ac872bec7b40910) Thanks [@yukukotani](https://github.com/yukukotani)! - Add support for size token

- [#291](https://github.com/kuma-ui/kuma-ui/pull/291) [`066d043`](https://github.com/kuma-ui/kuma-ui/commit/066d04383d3f4e19fb0460b184e8aa9ed67ae41d) Thanks [@yukukotani](https://github.com/yukukotani)! - Add support for radius token

### Patch Changes

- [#286](https://github.com/kuma-ui/kuma-ui/pull/286) [`144e1d0`](https://github.com/kuma-ui/kuma-ui/commit/144e1d01a852ca21fc4a89da7f5da2a1f0a4f955) Thanks [@MH4GF](https://github.com/MH4GF)! - chore: modified existing test cases with the introduction of testing-library/jest-dom

- Updated dependencies [[`bc1a7f9`](https://github.com/kuma-ui/kuma-ui/commit/bc1a7f979f5539a4e2d1d2785dc2a893eee076eb), [`10f9811`](https://github.com/kuma-ui/kuma-ui/commit/10f98111d542f1fa45a6cda18ac872bec7b40910), [`066d043`](https://github.com/kuma-ui/kuma-ui/commit/066d04383d3f4e19fb0460b184e8aa9ed67ae41d)]:
  - @kuma-ui/system@1.3.0
  - @kuma-ui/sheet@1.1.0

## 1.1.3

### Patch Changes

- [#276](https://github.com/kuma-ui/kuma-ui/pull/276) [`e276aea`](https://github.com/kuma-ui/kuma-ui/commit/e276aea4dbd933800de5237d0411df9097451dfe) Thanks [@naruaway](https://github.com/naruaway)! - fix: Replace broken generateKey with fast-json-stable-stringify for pseudo props

- [#278](https://github.com/kuma-ui/kuma-ui/pull/278) [`a03add5`](https://github.com/kuma-ui/kuma-ui/commit/a03add5b2a0b4ec9bcd6d2fd636c6d03a14f1636) Thanks [@MH4GF](https://github.com/MH4GF)! - maintenance: setup testing-library/jest-dom and add component test for Box

- [#272](https://github.com/kuma-ui/kuma-ui/pull/272) [`2329a1d`](https://github.com/kuma-ui/kuma-ui/commit/2329a1d9ea2fe4cf2d61cdf4ea87983c75b0d8fb) Thanks [@taishinaritomi](https://github.com/taishinaritomi)! - feat: support font theme tokens

- Updated dependencies [[`2329a1d`](https://github.com/kuma-ui/kuma-ui/commit/2329a1d9ea2fe4cf2d61cdf4ea87983c75b0d8fb)]:
  - @kuma-ui/sheet@1.0.2
  - @kuma-ui/system@1.2.1

## 1.1.2

### Patch Changes

- Updated dependencies [[`022fdec`](https://github.com/kuma-ui/kuma-ui/commit/022fdecfd1c2494974e9965ddca31aebf03d1dc2), [`90a9e65`](https://github.com/kuma-ui/kuma-ui/commit/90a9e6504d9e11f164f2373a8968e8eef7130b9d)]:
  - @kuma-ui/system@1.2.0

## 1.1.1

### Patch Changes

- Updated dependencies [[`8a42a22`](https://github.com/kuma-ui/kuma-ui/commit/8a42a225359bad6d2bf35a3c63d897ae02f59392)]:
  - @kuma-ui/system@1.1.0

## 1.1.0

### Minor Changes

- [#224](https://github.com/poteboy/kuma-ui/pull/224) [`b9110d7`](https://github.com/poteboy/kuma-ui/commit/b9110d7e17c7ad16b70604f2272adb53c96c5544) Thanks [@rabitarochan](https://github.com/rabitarochan)! - fix: resolving paths on Windows

## 1.0.1

### Patch Changes

- Updated dependencies [[`ee93574`](https://github.com/poteboy/kuma-ui/commit/ee9357493694118f6c9c829c9107221688a65ab8)]:
  - @kuma-ui/sheet@1.0.1
  - @kuma-ui/system@1.0.1

## 1.0.0

### Major Changes

- [`8ada2cd`](https://github.com/poteboy/kuma-ui/commit/8ada2cd64f144103cf611fc0990f4f74bcc19969) Thanks [@poteboy](https://github.com/poteboy)! - üêª‚Äç‚ùÑÔ∏è

### Patch Changes

- Updated dependencies [[`8ada2cd`](https://github.com/poteboy/kuma-ui/commit/8ada2cd64f144103cf611fc0990f4f74bcc19969)]:
  - @kuma-ui/sheet@1.0.0
  - @kuma-ui/system@1.0.0

## 0.6.0-next.1

### Minor Changes

- [#126](https://github.com/poteboy/kuma-ui/pull/126) [`c4b7318`](https://github.com/poteboy/kuma-ui/commit/c4b73184ca8e9c659734f9547143806a3a7d222a) Thanks [@poteboy](https://github.com/poteboy)! - Introducing Theming, Hybrid CSS-in-JS, and Universal UI Components

### Patch Changes

- Updated dependencies [[`c4b7318`](https://github.com/poteboy/kuma-ui/commit/c4b73184ca8e9c659734f9547143806a3a7d222a)]:
  - @kuma-ui/system@0.7.0-next.1
  - @kuma-ui/sheet@0.6.0-next.1

## 0.6.0-next.0

### Minor Changes

- [#126](https://github.com/poteboy/kuma-ui/pull/126) [`c89e97c`](https://github.com/poteboy/kuma-ui/commit/c89e97c2f2d3fc56586d0e02d5d8ff873b9eee6a) Thanks [@poteboy](https://github.com/poteboy)! - Introducing Theming, Hybrid CSS-in-JS, and Universal UI Components in @kuma-ui/core

### Patch Changes

- Updated dependencies [[`c89e97c`](https://github.com/poteboy/kuma-ui/commit/c89e97c2f2d3fc56586d0e02d5d8ff873b9eee6a)]:
  - @kuma-ui/system@0.7.0-next.0
  - @kuma-ui/sheet@0.6.0-next.0

## 0.5.0

### Minor Changes

- [#139](https://github.com/poteboy/kuma-ui/pull/139) [`6d0ba6d`](https://github.com/poteboy/kuma-ui/commit/6d0ba6db946f38de2bb3ea51b288daddfbb67196) Thanks [@poteboy](https://github.com/poteboy)! - Pre-release Bug Fix: Limit style props extraction to Kuma UI components

## 0.4.2

### Patch Changes

- 8af4f8d: chore: Automate synchronization of README.md across packages.
- Updated dependencies [8af4f8d]
  - @kuma-ui/system@0.6.2

## 0.4.1

### Patch Changes

- 558823e: fix classname hash
  - @kuma-ui/system@0.6.1

## 0.4.0

### Minor Changes

- Fully support RSC & Fully support pseudo elements & classes & throw an error when accessing any property of 'k' at runtime

### Patch Changes

- Updated dependencies
  - @kuma-ui/system@0.6.0

## 0.3.1

### Patch Changes

- @kuma-ui/system@0.5.1

## 0.3.0

### Minor Changes

- Documentation site

### Patch Changes

- Updated dependencies
  - @kuma-ui/system@0.5.0

## 0.2.0

### Minor Changes

- Added js-doc to css props type & Add css Function

## 0.1.0

### Minor Changes

- experimental support for Next.js 13.4 App Router and React Server Components

## 0.0.4

### Patch Changes

- Overriding styles defined within tagged template literals using style props

## 0.0.3

### Patch Changes

- Replace global whitespace removal regex with more specific regex for CSS properties #23
  Refactor HTML Tags. #24
  Fix incorrect setup example for Next.js #25
  fix position 'bottom, inset' css prop values #26

## 0.0.2

### Patch Changes

- initial release of kuma-ui packages
