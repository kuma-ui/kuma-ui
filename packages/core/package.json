{
  "name": "@kuma-ui/core",
  "version": "1.5.6",
  "description": "🐻 Kuma UI is a utility-first, zero-runtime CSS-in-JS library that offers an outstanding developer experience and optimized performance.",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/kuma-ui/kuma-ui.git"
  },
  "keywords": [
    "react",
    "css-in-js",
    "zero-runtime",
    "front-end",
    "design-system",
    "styled-components"
  ],
  "homepage": "https://www.kuma-ui.com",
  "main": "dist/index.js",
  "module": "dist/index.mjs",
  "types": "dist/index.d.ts",
  "sideEffects": false,
  "exports": {
    ".": {
      "types": "./dist/index.d.ts",
      "require": "./dist/index.js",
      "import": "./dist/index.mjs"
    },
    "./components/componentList": {
      "types": "./dist/components/componentList.d.ts",
      "require": "./dist/components/componentList.js",
      "import": "./dist/components/componentList.mjs"
    }
  },
  "typesVersions": {
    "*": {
      "components/componentList": [
        "./dist/components/componentList.d.ts"
      ]
    }
  },
  "files": [
    "dist"
  ],
  "scripts": {
    "build": "tsup",
    "test": "vitest run && vitest typecheck --run",
    "typecheck": "tsc --noEmit --composite false",
    "lint": "eslint './src/**/*.{js,ts,jsx,tsx}' --report-unused-disable-directives --max-warnings 0",
    "lint:fix": "eslint --fix './src/**/*.{js,ts,jsx,tsx}' --report-unused-disable-directives --max-warnings 0"
  },
  "dependencies": {
    "@kuma-ui/sheet": "workspace:*",
    "@kuma-ui/system": "workspace:*",
    "@types/stylis": "^4.2.0",
    "stylis": "^4.2.0"
  },
  "devDependencies": {
    "@testing-library/jest-dom": "^5.16.5",
    "@testing-library/react": "^14.0.0",
    "@types/react": "^18.0.32",
    "@types/testing-library__jest-dom": "^5.14.9",
    "@vitest/coverage-c8": "^0.31.4",
    "globby": "^13.2.0",
    "jsdom": "^22.1.0",
    "ts-node": "^10.9.1",
    "vitest": "^0.31.4"
  },
  "peerDependencies": {
    "@types/react": "^18.0.32",
    "next": ">=13.4.5",
    "react": ">=18.2.0"
  },
  "peerDependenciesMeta": {
    "@types/react": {
      "optional": true
    },
    "next": {
      "optional": true
    }
  },
  "publishConfig": {
    "access": "public"
  }
}
