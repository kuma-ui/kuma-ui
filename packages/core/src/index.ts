export { styled } from "./styled";
export { k } from "./k";
export { css, cx } from "./css";
export { createTheme, type Theme, type ThemeSystem } from "./theme";
export * from "./components";
export * from "./registry";
