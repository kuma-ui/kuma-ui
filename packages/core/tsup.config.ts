import { defineConfig } from "tsup";
import fs from "fs";
import path from "path";

export default defineConfig([
  {
    clean: true,
    dts: true,
    format: ["cjs"],
    entry: [
      "src",
      "!src/**/*.test.*",
      "!src/**__test__/**",
      "!src/**/vitest.setup.ts",
    ],
    bundle: false,
  },
  {
    clean: true,
    format: ["esm"],
    entry: [
      "src",
      "!src/**/*.test.*",
      "!src/**__test__/**",
      "!src/**/vitest.setup.ts",
    ],
    bundle: true,
    esbuildPlugins: [
      {
        name: "add-extension",
        setup(build) {
          build.onResolve({ filter: /.*/ }, (args) => {
            const fileExtension = ".ts";
            const extension = ".mjs";
            if (args.importer) {
              const p = path.join(args.resolveDir, args.path);
              let tsPath = `${p}${fileExtension}`;
              let tsxPath = `${p}${fileExtension}x`;

              let importPath = "";
              if (fs.existsSync(tsPath) || fs.existsSync(tsxPath)) {
                importPath = args.path + extension;
              } else {
                tsPath = path.join(
                  args.resolveDir,
                  args.path,
                  `index${fileExtension}`,
                );
                tsxPath = path.join(
                  args.resolveDir,
                  args.path,
                  `index${fileExtension}x`,
                );
                if (fs.existsSync(tsPath) || fs.existsSync(tsxPath)) {
                  importPath = `${args.path}/index${extension}`;
                }
              }
              return { path: importPath, external: true };
            }
          });
        },
      },
    ],
  },
]);
