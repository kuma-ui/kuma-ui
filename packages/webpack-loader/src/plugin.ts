import type { Compilation, Compiler, NormalModule } from "webpack";
import { theme, sheet } from "@kuma-ui/sheet";
import { transform } from "@kuma-ui/babel-plugin";
import path from "path";
import { RawSource } from "webpack-sources";
import kumaUiLoader from "./loader";

type WebpackPluginOption = {
  breakpoints?: Record<string, string>; // {sm: '400px', md: '700px'}
};

const pluginName = "KumaUIWebpackPlugin";

class KumaUIWebpackPlugin {
  options: WebpackPluginOption;

  constructor({ options = {} }) {
    this.options = options;
    if (
      this.options.breakpoints &&
      Object.keys(this.options.breakpoints).length > 0
    ) {
      theme.setBreakpoints(this.options.breakpoints);
    }
  }

  apply(compiler: Compiler) {
    compiler.hooks.afterCompile.tap(pluginName, (compilation: Compilation) => {
      const cssLookup: { [key: string]: string } = {};
      compilation.hooks.succeedModule.tap(pluginName, async (module) => {
        if ("resource" in module && typeof module.resource === "string") {
          const id = module.resource;
          if (
            id.includes("/node_modules/") ||
            id.includes("@kuma-ui/core") ||
            !/\.(t|j)(s|sx)?$/.test(id)
          ) {
            return;
          }
          if ("_source" in module) {
            const originalCode: string = (module._source as any).source();
            const result = await transform(originalCode, id);
            if (!result || !result.code) return;
            const cssFilename = path.normalize(
              `${id.replace(/\.[jt]sx?$/, "")}.css`
            );
            const cssRelativePath = path
              .relative(process.cwd(), cssFilename)
              .replace(/\\/g, path.posix.sep);
            const cssId = `/${cssRelativePath}`;
            const css = sheet.getCSS();
            cssLookup[cssFilename] = css;
            cssLookup[cssId] = css;
            console.log(cssLookup);
            console.log(result.code);
            module._source = new RawSource(result.code);
          }
        }
      });
    });
  }
}

export default KumaUIWebpackPlugin;
