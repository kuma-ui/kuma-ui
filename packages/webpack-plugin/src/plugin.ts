import { Compiler } from "webpack";
import { theme } from "@kuma-ui/sheet";
import path from "path";
import { readdirSync, writeFileSync, existsSync } from "fs";
import { getUserTheme } from "./getUserTheme";
import { createRequire } from "module";
import { fileURLToPath } from "node:url";

// tsup will replace __ESM__ with true during ESM build and false during CJS build when bundling.
declare const __ESM__: boolean;

const _require = __ESM__ ? createRequire(import.meta.url) : require;

const _filename = __ESM__ ? fileURLToPath(import.meta.url) : __filename;
const _dirname = __ESM__ ? path.dirname(_filename) : __dirname;

export const CSS_PATH = (() => {
  const cssPath = path.join(_dirname, "..", "assets", "kuma.css");
  if (!existsSync(cssPath)) writeFileSync(cssPath, "");
  return cssPath;
})();

type KumaWebpackOptions = {
  /** The destination to emit an actual CSS file. This is a temporary workaround to enable HMR in Client Component @see loader.ts */
  outputDir?: string;
};

class KumaUIWebpackPlugin {
  /* @internal */
  config: string | undefined;
  /* @internal */
  watchMode = false;

  private outputDir: string | undefined;

  constructor(options: KumaWebpackOptions = {}) {
    this.outputDir = options.outputDir;

    const dir = readdirSync(".");
    dir.forEach((filePath) => {
      if (filePath.startsWith("kuma.config.")) {
        this.config = path.join(process.cwd(), filePath);
      }
    });
  }

  apply(compiler: Compiler) {
    const { config } = this;

    compiler.options.plugins.push(
      new compiler.webpack.DefinePlugin({
        ...(config
          ? {
              "globalThis.__KUMA_USER_THEME__":
                compiler.webpack.DefinePlugin.runtimeValue(
                  () => {
                    const userTheme = getUserTheme(config);
                    if (userTheme) {
                      theme.setUserTheme(userTheme);
                    }
                    return JSON.stringify(userTheme);
                  },
                  {
                    // enable automatic rebuild for dynamic theme props
                    // <Box color={(() => "colors.red.100"))()} />
                    fileDependencies: [config],
                  },
                ),
            }
          : undefined),
      }),
    );

    compiler.hooks.watchRun.tap("@kuma-ui/webpack-plugin:watchRun", () => {
      this.watchMode = true;
    });

    compiler.options.module?.rules?.push(
      {
        test: /\.(tsx|ts|js|mjs|jsx)$/,
        exclude: /node_modules/,
        use: [
          {
            loader: _require.resolve("./loader.js"),
            options: { plugin: this, outputDir: this.outputDir },
          },
        ],
      },
      {
        test: CSS_PATH,
        use: [
          {
            loader: _require.resolve("./cssLoader.js"),
            options: { plugin: this },
          },
        ],
      },
    );
  }
}

export default KumaUIWebpackPlugin;
