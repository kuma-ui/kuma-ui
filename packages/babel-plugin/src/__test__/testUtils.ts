import { BabelFileResult, transformSync } from "@babel/core";
import plugin from "../";
import { types, template } from "@babel/core";

export function babelTransform(
  code: string,
  runtime: "classic" | "automatic" = "classic"
) {
  const filename = "test.tsx";
  let result: BabelFileResult | null = null;
  result = transformSync(code, {
    presets: [
      "@babel/preset-typescript",
      [
        "@babel/preset-react",
        {
          runtime,
        },
      ],
    ],
    filename,
  });

  if (result === null || result.code == null)
    throw new Error(`Could not transform`);

  result = transformSync(result.code, {
    plugins: [plugin({ types, template })],
    filename,
  });

  if (result === null || result.code == null)
    throw new Error(`Could not transform`);

  return result;
}

export function getExpectSnapshot(result: BabelFileResult) {
  return `
${(result.metadata as { css: string } | undefined)?.css}

${result.code}
`;
}
