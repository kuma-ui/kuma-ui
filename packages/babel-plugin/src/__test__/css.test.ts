// for babel-plugin-tester; see: https://github.com/babel-utils/babel-plugin-tester#vitest
/// <reference types="vitest/globals" />

import { babelTransform } from "./testUtils";
import { pluginTester } from "babel-plugin-tester";
import { types, template } from "@babel/core";
import plugin from "../";
import path from "path";

describe("css function", () => {
  describe("Snapshot tests", () => {
    test("basic usage should match snapshot", () => {
      // Arrange
      const inputCode = `
        import { css } from '@kuma-ui/core'
        const style = css({ color: 'red' })
      `;
      // Act
      const { code } = babelTransform(inputCode);
      // Assert
      expect(code).toMatchSnapshot();
    });

    test("using space props should match snapshot", () => {
      // Arrange
      const inputCode = `
        import { css } from '@kuma-ui/core'
        const style = css({ p: 2 })
      `;
      // Act
      const { code } = babelTransform(inputCode);
      // Assert
      expect(code).toMatchSnapshot();
    });

    test("using pseudo props should match snapshot", () => {
      // Arrange
      const inputCode = `
        import { css } from '@kuma-ui/core'
        const style = css({ _hover: { color: 'red' } })
      `;
      // Act
      const { code } = babelTransform(inputCode);
      // Assert
      expect(code).toMatchSnapshot();
    });
  });
});

pluginTester({
  plugin: () => plugin({ types, template }),
  babelOptions: {
    presets: [
      "@babel/preset-typescript",
      [
        "@babel/preset-react",
        {
          runtime: "classic",
        },
      ],
    ],
  },
  filename: "test.tsx",
  fixtures: path.join(__dirname, "__fixtures__"),
});
