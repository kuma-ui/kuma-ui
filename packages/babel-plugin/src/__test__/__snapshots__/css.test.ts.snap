// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`css function > Snapshot tests > basic usage should match snapshot 1`] = `

.kuma-3383855968 {color:red;}

import React from "react";
import { css } from '@kuma-ui/core';
const style = "kuma-3383855968";

`;

exports[`css function > Snapshot tests > using pseudo elements should match snapshot 1`] = `

.kuma-1234644783 {}.kuma-1234644783:after{color:blue;}

import React from "react";
import { css } from '@kuma-ui/core';
const style = "kuma-1234644783";

`;

exports[`css function > Snapshot tests > using pseudo props should match snapshot 1`] = `

.kuma-1234644783 {}.kuma-1234644783:hover{color:red;}

import React from "react";
import { css } from '@kuma-ui/core';
const style = "kuma-1234644783";

`;

exports[`css function > Snapshot tests > using space props should match snapshot 1`] = `

.kuma-110275621 {padding:2px;}

import React from "react";
import { css } from '@kuma-ui/core';
const style = "kuma-110275621";

`;
